library/core/src/iter/traits/iterator.rs

```diff
@@ -48,13 +48,24 @@ pub trait Iterator {
     type Item;
 
     /// Advances the iterator and returns the next value.
+    /// Returns [`None`] when iteration is finished.
     ///
-    /// Returns [`None`] when iteration is finished. Individual iterator
-    /// implementations may choose to resume iteration, and so calling `next()`
-    /// again may or may not eventually start returning [`Some(Item)`] again at some
-    /// point.
+    /// # Behavior after [`None`] is returned
+    ///
+    /// Calling `next()` on an iterator after it has returned `None`
+    /// is in general a perfectly valid use of the interface.
+    /// Implementations may choose to resume iteration, and so calling `next()`
+    /// again may eventually start returning [`Some(Item)`] again at some
+    /// point. Alternately, implementations may choose to continue returning [`None`]
+    /// forever, in which case they should implement [`FusedIterator`] as well.
+    ///
+    /// Implementations should document their behavior, or their lack of a guarantee
+    /// about their behavior, after `None` is returned. If they must panic,
```

Comment:
Maybe this should get a carve-out for iterators with FusedIterator impls. A FusedIterator impl is as good as a boilerplate sentence that says "after returning None I will continue returning None" and we shouldn't require or recommend both.